src/node_zip.cc

```diff
@@ -0,0 +1,454 @@
+
+  zip_int64_t num_entries = zip_get_num_entries(zip->zip_, 0);
+  v8::Local<v8::Array> entries = v8::Array::New(isolate, num_entries);
+
+  for (zip_int64_t t = 0; t < num_entries; ++t) {
+    char const* entry_name = zip_get_name(zip->zip_, t, 0);
+    CHECK_NOT_NULL(entry_name);
+
+    v8::Local<v8::Array> entry = v8::Array::New(isolate, 2);
+    entry->Set(ctx, v8::Integer::New(isolate, 0), v8::String::NewFromUtf8(isolate, entry_name).ToLocalChecked());
+    entry->Set(ctx, v8::Integer::New(isolate, 1), v8::Integer::New(isolate, with_file_types ? zip->GetFileType(t) : 0));
+
+    entries->Set(ctx, v8::Integer::New(isolate, t), entry);
+  }
+
+  args.GetReturnValue().Set(entries);
+}
+
+void ZipArchive::AddEntry(const v8::FunctionCallbackInfo<v8::Value>& args) {
+  v8::Isolate* isolate = args.GetIsolate();
+
+  ZipArchive* zip;
+  ZipArchive::CheckZip(args, zip);
+
+  const int argc = args.Length();
+  CHECK_EQ(argc, 5);
+
+  CHECK(args[0]->IsString());
+  node::Utf8Value path(isolate, args[0]);
+
+  CHECK(args[1]->IsInt32());
+  zip_int32_t comp_method = args[1].As<v8::Int32>()->Value();
+
+  CHECK(Buffer::HasInstance(args[2]));
+  v8::Local<v8::Object> buffer_obj = args[2].As<v8::Object>();
+  char* buffer_data = Buffer::Data(buffer_obj);
+  size_t buffer_length = Buffer::Length(buffer_obj);
+
+  void* buffer_copy = malloc(buffer_length);
+  memcpy(buffer_copy, buffer_data, buffer_length);
+
+  zip_source_t* file_source = zip_source_buffer(zip->zip_, buffer_copy, buffer_length, 1);
+  CHECK_NOT_NULL(file_source);
+
+  if (comp_method != ZIP_CM_STORE) {
+    CHECK(args[3]->IsUint32());
+    zip_uint32_t crc = args[3].As<v8::Uint32>()->Value();
+
+    CHECK(args[4]->IsUint32());
+    zip_uint32_t size = args[4].As<v8::Uint32>()->Value();
+
+    std::cout << "set comp = " << comp_method << " / crc = " << crc << " / size = " << size << std::endl;
+
+    CHECK_EQ(zip_source_buffer_set_compression(file_source, comp_method, size, crc), 0);
+  }
+
+  zip_int64_t file_index = zip_file_add(zip->zip_, *path, file_source, ZIP_FL_OVERWRITE | ZIP_FL_ENC_UTF_8);
+  CHECK_GE(file_index, 0);
+  CHECK_LE(file_index, std::numeric_limits<uint32_t>::max());
+
+  args.GetReturnValue().Set(static_cast<uint32_t>(file_index));
+}
+
+void ZipArchive::CheckZip(const v8::FunctionCallbackInfo<v8::Value>& args, ZipArchive*& zip) {
+  ASSIGN_OR_RETURN_UNWRAP(&zip, args.Holder());
+  CHECK_NOT_NULL(zip->zip_);
+  CHECK_NOT_NULL(zip->source_);
+}
+
+void ZipArchive::CheckZipEntry(const v8::FunctionCallbackInfo<v8::Value>& args, ZipArchive*& zip, zip_uint32_t& index) {
+  v8::Isolate* isolate = args.GetIsolate();
+  Environment* env = Environment::GetCurrent(isolate);
+  v8::Local<v8::Context> ctx = env->context();
+
+  ASSIGN_OR_RETURN_UNWRAP(&zip, args.Holder());
+  CHECK_NOT_NULL(zip->zip_);
+  CHECK_NOT_NULL(zip->source_);
+
+  const int argc = args.Length();
+  CHECK_GE(argc, 1);
+
+  CHECK(args[0]->Uint32Value(ctx).To(&index));
```

Comment:
more idiomatic (this would apply elsewhere in this PR too):

```suggestion
  CHECK(args[0]->IsUint32());
  index = args[0].As<Uint32>()->Value();
```

or (with error handling, in which case you'd return a `v8::Maybe`):

```suggestion
  if (!args[0]->Uint32Value(ctx).To(&index)) return {};
```